### PR TITLE
Use atomics for StartStopOnce implementation

### PR DESCRIPTION
**Pull request by archseer**

### CODE REVIEW

File: core/utils/utils.go [hunk 1]
```diff
@@ -866,65 +866,47 @@ const (
...
-	defer once.Unlock()
+	success := once.state.CAS(int32(StartStopOnce_Unstarted), int32(StartStopOnce_Started))
 
-	if once.state != StartStopOnce_Unstarted {
+	if !success {
 		return errors.Errorf("%v has already started once", name)
 	}
-	once.state = StartStopOnce_Started
 
 	return fn()
```
Comment (archseer) [inline core/utils/utils.go:875]: Hmm, might not be ideal if stop is called while the function is still executing. I could keep a `Mutex` specifically for `Start`/`Stop` calls (locking after the `CAS`), that way at least `IfStarted`/`Status` doesn't need to hold a lock.

Comment (archseer) [reply core/utils/utils.go:879]: I've done that, though I'm not sure if it's really necessary in practice.

File: core/utils/utils.go [hunk 2]
```diff
@@ -866,65 +867,53 @@ const (
...
 // StopOnce sets the state to Stopped
 func (once *StartStopOnce) StopOnce(name string, fn func() error) error {
-	once.Lock()
-	defer once.Unlock()
+	success := once.state.CAS(int32(StartStopOnce_Started), int32(StartStopOnce_Stopped))
 
-	if once.state != StartStopOnce_Started {
-		return errors.Errorf("%v has already stopped once", name)
+	if !success {
+		return errors.Errorf("%v has already started once", name)
```
Comment (PiotrTrzpil) [inline core/utils/utils.go:887]: i think the original 'stopped' is correct here?

File: core/utils/utils.go [hunk 3]
```diff
@@ -866,65 +867,53 @@ const (
...
 	}
-	once.state = StartStopOnce_Stopped
+
+	once.Lock()
+	defer once.Unlock()
 
 	return fn()
 }
 
 // OkayToStart checks if the state may be started
```
Comment (PiotrTrzpil) [inline core/utils/utils.go:896]: minor, but the comment seems outdated, as it not only checks, but also changes the state. Similar for OkayToStop

**Review by samsondav — COMMENTED** [review]
What does this fix?

File: core/utils/utils.go [hunk 4]
```diff
@@ -866,65 +867,53 @@ const (
 
 // StartOnce sets the state to Started
 func (once *StartStopOnce) StartOnce(name string, fn func() error) error {
-	once.Lock()
-	defer once.Unlock()
+	success := once.state.CAS(int32(StartStopOnce_Unstarted), int32(StartStopOnce_Started))
```
Comment (j16r) [inline core/utils/utils.go:870]: Could we just do something like:

```
func (once *StartStopOnce) StartOnce() {
ran := false
sync.Once(func() {
  ran = true
  once.started = true
}
if ran == false {
  return errors.New("already started")
}

})

**Comment by archseer** [comment]
Oh also, the benefit is that calling `Start()` repeatedly concurrently won't block anymore but instead immediately return an error

File: core/utils/utils.go (hunk 4 above)
Comment (archseer) [reply core/utils/utils.go:870]: `once.started` would need to be an atomic in this case or it's unsafe to read from. `fn` also needs to be called while the lock is held, or we can call `StopOnce` before `StartOnce` finishes executing.

**Comment by archseer** [comment]
Addressed review, it's ready for another look.

**Comment by archseer** [comment]
Merging the changes from here into a new PR

**Pull request closed without merging**